js/src/util/tokens/index.js

```diff
@@ -0,0 +1,293 @@
+    .then((results) => {
+      return results.map((rawImage) => {
+        const image = tokenReg.instance.meta.decodeOutput(rawImage)[0].value;
+
+        return hashToImageUrl(image);
+      });
+    });
+}
+
+/**
+ * `updates` should be in the shape:
+ *   {
+ *     [ who ]: [ tokenId ]  // Array of tokens to updates
+ *   }
+ *
+ * Returns a Promise resolved with the balances in the shape:
+ *   {
+ *     [ who ]: { [ tokenId ]: BigNumber } // The balances of `who`
+ *   }
+ */
+export function fetchAccountsBalances (api, tokens, updates) {
+  const accountAddresses = Object.keys(updates);
+
+  // Updates for the ETH balances
+  const ethUpdates = accountAddresses
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId === ETH_TOKEN.id);
+    })
+    .reduce((nextUpdates, tokenIds, accountIndex) => {
+      if (tokenIds.length > 0) {
+        const accountAddress = accountAddresses[accountIndex];
+
+        nextUpdates[accountAddress] = tokenIds;
+      }
+
+      return nextUpdates;
+    }, {});
+
+  // Updates for Tokens balances
+  const tokenUpdates = Object.keys(updates)
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId !== ETH_TOKEN.id);
+    })
+    .reduce((nextUpdates, tokenIds, accountIndex) => {
+      if (tokenIds.length > 0) {
+        const accountAddress = accountAddresses[accountIndex];
+
+        nextUpdates[accountAddress] = tokenIds;
+      }
+
+      return nextUpdates;
+    }, {});
+
+  let ethBalances = {};
+  let tokensBalances = {};
+
+  const ethPromise = fetchEthBalances(api, Object.keys(ethUpdates))
+    .then((_ethBalances) => {
+      ethBalances = _ethBalances;
+    });
+
+  let tokenPromise = Promise.resolve();
+
+  Object.keys(tokenUpdates)
+    .forEach((accountAddress) => {
+      const tokenIds = tokenUpdates[accountAddress];
+      const updateTokens = tokens
+        .filter((t) => tokenIds.includes(t.id));
+
+      tokenPromise = tokenPromise
+        .then(() => fetchTokensBalances(api, updateTokens, [ accountAddress ]))
+        .then((balances) => {
+          tokensBalances[accountAddress] = balances[accountAddress];
+        });
+    });
+
+  return ethPromise
+    .then(() => tokenPromise)
+    .then(() => {
+      const balances = Object.assign({}, tokensBalances);
+
+      Object.keys(ethBalances).forEach((accountAddress) => {
+        if (!balances[accountAddress]) {
+          balances[accountAddress] = {};
+        }
+
+        balances[accountAddress] = Object.assign(
+          {},
+          balances[accountAddress],
+          ethBalances[accountAddress]
+        );
+      });
+
+      return balances;
+    });
+}
+
+function fetchEthBalances (api, accountAddresses) {
+  const promises = accountAddresses
+    .map((accountAddress) => api.eth.getBalance(accountAddress));
+
+  return Promise.all(promises)
+    .then((balancesArray) => {
+      return balancesArray.reduce((balances, balance, index) => {
```

Comment:
In the future it might actually be better to display balances as they come instead of waiting for all of them.

Reply:
Yeah, right now it's first the ETH balances, then tokens. But with Redux, the more we dispatch a new state, the more it re-renders a *lot* of Components...

Reply:
So I spoke/wrote too fast. It shouldn't re-render components, but will run the dispatcher methods and compare the results for every rendered components, so it's anyway better to do it once than multiple times.